mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp

```diff
@@ -772,6 +787,964 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<GridwiseGemmOp> {
+      attentionOutAccBufferMaxScaled = rewriter.create<arith::MulFOp>(
+          loc, attentionOutAccBufferMaxScaled, ldAttentionOutAccBuffer);
+
+      // attentionOutAccBufferMaxScaledNoSumDiv = attentionOutAccBufferMaxScaled
+      // * sumRowBuffer
+      // --------------------------------------------------------------------
+      Value attentionOutAccBufferMaxScaledNoSumDiv =
+          rewriter.create<arith::MulFOp>(loc, attentionOutAccBufferMaxScaled,
+                                         ldSumRowBuffer);
+
+      // gemm1OutThreadwiseViewMaxScaled  = exp(gemm0OutBufferMaxView[:,0] -
+      // maxRowBufferNew) * gemm1OutThreadwiseView
+      // --------------------------------------------------------------------
+      Value gemm1OutThreadwiseViewMaxScaled = rewriter.create<arith::SubFOp>(
+          loc, ldgemm0OutBufferMax, maxRowBufferNew);
+      gemm1OutThreadwiseViewMaxScaled =
+          rewriter.create<math::ExpOp>(loc, gemm1OutThreadwiseViewMaxScaled);
+      Value ldGemm1Out = rewriter.create<InBoundsLoadOp>(
+          loc, outElemType, gemm1Out, gemm1OutCoords);
+      gemm1OutThreadwiseViewMaxScaled = rewriter.create<arith::MulFOp>(
+          loc, gemm1OutThreadwiseViewMaxScaled, ldGemm1Out);
+
+      // [STORE] attentionOutAccBuffer = (attentionOutAccBufferMaxScaledNoSumDiv
+      // + gemm1OutThreadwiseViewMaxScaled ) / sumRowBufferNew
+      // --------------------------------------------------------------------
+      Value stAttentionOutAccBuffer = rewriter.create<arith::AddFOp>(
+          loc, attentionOutAccBufferMaxScaledNoSumDiv,
+          gemm1OutThreadwiseViewMaxScaled);
+      stAttentionOutAccBuffer = rewriter.create<arith::DivFOp>(
+          loc, stAttentionOutAccBuffer, sumRowBufferNew);
+      rewriter.create<InBoundsStoreOp>(loc, stAttentionOutAccBuffer,
+                                       attentionOutAccBuffer,
+                                       attentionOutAccBufferCoords);
+
+      // [STORE] maxRowBuffer = maxRowBufferNew
+      // [STORE] sumRowBuffer = sumRowBufferNew
+      // --------------------------------------------------------------------
+      // We only need to do with once per row
+      Value isNptLastIdx = rewriter.create<arith::CmpIOp>(
+          loc, arith::CmpIPredicate::eq, upperCoords[1], lastNptIdx);
+      scf::IfOp ifb = rewriter.create<scf::IfOp>(loc, isNptLastIdx,
+                                                 /*withElseRegion=*/false);
+      {
+        OpBuilder thenb = ifb.getThenBodyBuilder();
+        thenb.create<InBoundsStoreOp>(loc, maxRowBufferNew, maxRowBuffer,
+                                      ValueRange{upperCoords[0]});
+        thenb.create<InBoundsStoreOp>(loc, sumRowBufferNew, sumRowBuffer,
+                                      ValueRange{upperCoords[0]});
+      }
+    }
+  }
+
+  RockAccelTuningParamAttrInterface
+  deriveGemm1TuningParams(PatternRewriter &rewriter,
```

Comment:
Why's this function down here?

Reply:
My reasons are that it is an implementation detail of gridwise attention lowering and not a real tuning parameter. Therefore should not be exposed before this stage of lowering.
There is no real freedom of choice for the values unlike gemm0 tuning param -- not an heuristic